src/libcore/str/pattern.rs

```diff
@@ -706,8 +708,16 @@ unsafe impl<'a, 'b> ReverseSearcher<'a> for StrSearcher<'a, 'b> {
                 }
             }
             StrSearcherImpl::TwoWay(ref mut searcher) => {
-                searcher.next_back::<MatchOnly>(self.haystack.as_bytes(),
-                                                self.needle.as_bytes())
+                let is_long = searcher.memory == usize::MAX;
+                if is_long {
+                    searcher.next_back::<MatchOnly>(self.haystack.as_bytes(),
+                                                    self.needle.as_bytes(),
+                                                    true)
```

Comment:
I'm unsure if this is actually working to compile to two versions. Wrapping Long vs Short period cases into the strategy trait could force it instead.

This was an idea from the original code, commit 39cb5b13e68b4616a1f1e8dff4fd3f5241e1842a